hamilton/experimental/h_async.py

```diff
@@ -0,0 +1,151 @@
+import asyncio
+import inspect
+import logging
+import types
+import typing
+from typing import Dict, Any, Type, Optional
+
+from hamilton import base, node, driver
+
+logger = logging.getLogger(__name__)
+
+
+async def await_dict_of_tasks(task_dict: Dict[str, types.CoroutineType]) -> Dict[str, Any]:
+    """Util to await a dictionary of tasks as asyncio.gather is kind of garbage"""
+    keys = sorted(task_dict.keys())
+    coroutines = [task_dict[key] for key in keys]
+    coroutines_gathered = await asyncio.gather(*coroutines)
+    return dict(zip(keys, coroutines_gathered))
+
+
+async def process_value(val: Any) -> Any:
+    """Helper function to process the value of a potential awaitable.
+    This is very simple -- all it does is await the value if its not already resolved.
+
+    @param val: Value to process.
+    @return: The value (awaited if it is a coroutine, raw otherwise).
+    """
+    if not inspect.isawaitable(val):
+        return val
+    return await val
+
+
+class AsyncGraphAdapter(base.SimplePythonDataFrameGraphAdapter):
+
+    def __init__(self, result_builder: base.ResultMixin = None):
+        """Creates an AsyncGraphAdapter class. Note this will *only* work with the AsyncDriver class.
+
+        Some things to note:
+        1. This executes everything at the end (recursively). E.G. the final DAG nodes are awaited
+        2. This does *not* work with decorators when the async function is being decorated. That is
+        because that function is called directly within the decorator, so we cannot await it.
+        """
+        super(AsyncGraphAdapter, self).__init__()
+        self.result_builder = result_builder if result_builder else base.PandasDataFrameResult()
+
+    @staticmethod
+    def check_input_type(node_type: Type, input_value: Any) -> bool:
+        """Checks the input type to ensure that it is compatible with the node type.
+        Note that this also allows coroutines, in case you want the DAG to await
+        the result of calling the input.
+
+        @param node_type: Type of the input node, expected
+        @param input_value: Type of the actual input, received
+        @return: Whether the actual input value matches the type expected
```

Comment:
we don't use this style :)
```suggestion
        :param node_type: Type of the input node, expected
        :param input_value: Type of the actual input, received
        :return: Whether the actual input value matches the type expected
```

Reply:
Ughh need to update my pycharm to auto-correct to the right one